### DISCUSSION TITLE
F´ GDS performance w.r.t real-time data visualization/monitoring

### DISCUSSION THREAD

**Discussion by ThibFrgsGmz** [nasa/fprime#1269] [Q&A] [answered]
Hello,

For a space payload demonstration project, we are considering using F´ but we would be quickly limited by the available functionalities of the F´ GDS. Indeed, our GDS would require for example:

- a continuous image display (>=30Hz);
- a telemetry grapher to display time tracking curves of specific data such as memory consumption, voltages or quaternions.

What strategy should we adopt?

Should we look to modify ``fprime-gds`` to get these measurement visualizations?

Or should we work immediately to integrate a third-party open source GDS framework to replace fprime-gds?
If so, do you have a recommendation on which GDS framework we should consider?

I have heard of ``OpenMCT``, ``Yamcs``, ``COSMOS``, ``Grafana``.

The choice of ``COSMOS`` doesn't seem very feasible to me because the latest version requires ``Docker`` (and also it's done in ``Ruby``, a language we don't know).

As far as ``OpenMCT`` is concerned, from what I've heard, ``OpenMCT`` will not natively support sending commands and will instead function as a data viewer similar to Grafana.
So I'm not sure if this integration/interface is feasible/simple to add the command sending plugin to the dashboard.

**Comment by LeStarch** [comment, marked as answer]
fprime-gds has the primary goal of supporting developers with a functional GDS for system testing early in project development. Thus it focuses on ease of use and ease of integration with F´ specifically.  It does have graphing capabilities (run up to 10Hz), and is somewhat extensible.  However it lacks some features of operations-focused GDSs, like persistent history, replay, etc.  Projects have adapted the fprime-gds to work with their operations model in the past, however; work had to be done to add in any critical missing features.

For example, some projects adapt the fprime-gds data flow, but add in a database for persisting history.  Some also adapt the GDS UI directly.

To make the decision of what GDS to use and how to adapt to it, I usually see the following questions:

1. What does the project, or the communications provider expect you to use?
2. What does the hardware expect you to use?
3. What do your operators expect to use?
4. What are the concept of operations of your project and how do you meet those requirements?

To dig in a little deeper, question 1 speaks to the entity (company, organization, etc) that will be getting your data to the ground.  Often there is an expectation from that entity that you will use a specific packet format or a specific GDS or both.  If this is the case, then you are likely adapting the ground interface layer inside F´flight code to match and then you should ask how to meet your operational requirements within that directive.  If you are provided "just a data pipe" then we move onto question 2.

Question 2 asks about your hardware.  Are there any bus, radio, protocol, or other requirements driven by the hardware that lend themselves to a particular format or GDS.  Remember, F´ and fprime-gds speak in F´ packets and if your hardware or required protocols preclude F´ packets, then you are likely adapting to that protocol on the FSW side as well.  However, some protocols wrap "just bytes" and some projects are happy with F´  packets so we move onto question 3 and 4.

Question 3 and 4 get to the heart of the concept of operations.  Are your operators requiring a specific GDS and does any GDS meet your requirements?  In this case, I am assuming you are free to recommend a GDS since you are asking about a route forward with F´so we will focus on question 4 with specifics.

The fprime-gds has a limited graphing ability (fixed time window) and has some extensibility in that you can develop custom dashboards and plugins. As mentioned before, it is primarily focused on in-lab testing.  Thus an understanding of your con-ops is required.  Is this a short mission or do you only care about recent data?  If so, you may be able to adapt the fprime-gds.  However, if you need strong historical data or recall you may wish to adapt the gds data pipelines to publish to some other GDS that meets your needs.

Some notes on the other GDSs that were mentioned:

**OpenMCT:**  quite a while back we were able to get F´  data published into OpenMCT but the lack of a commanding solution prevented us from adopting it in place of the fprime-gds.  We did not wish to maintain a custom commanding solution to integrate with OpenMCT and so we stayed the course with fprime-gds.  You theoretically could use OpenMCT as a telemetry store and graphing inteface, while retaining the fprime-gds commanding and sequencing tabs to do the commanding.  If I am remembering correctly, OpenMCT did not have the concept of Events either....so this may be an additional complication.

The OpenMCT work is found here: https://github.com/fprime-community/fprime-openmct-integration however, it is not something F´ is currently supporting.  It is there as a reference only.

**Cosmos:**  we also did an integration with Cosmos and had similar results to OpenMCT.  There were not events, but we did manage to attach Cosmos to F´ and have a similar reference here: https://github.com/fprime-community/fprime-cosmos-integration  Again, this is not something we are supporting and is for reference only.  This work is commingled with an old version of F´ and may not prove useful.

The others you mentioned we don't have direct experience with.

**Reply by LeStarch** [reply]
To speak to your specific questions, and my opinion therein I would advise this:

1. Ensure you can use the F´ packet format.  Otherwise you'll have to start this process with a completely separate packet/protocol and none of the advice apply.
2. Start with the 30Hz images and what GDSs support this....this is essentially live video and is a tall order
3. Subset to  GDSs that support live telemetry graphing as is your second requirement.
4. Then figure out how to get missing features (e.g. commanding)

fprime-gds might handle the live graphing to your satisfaction as long as the graphs are "recent data" and not long-running or historical data.  But my instinct is that it will be fairly difficult to get live 30Hz imagery updates.  If you go with something else to display telemetry, you likely should adapt the F´  ground components to interact with that GDS.  You may also be able to use the fprime-gds alongside the data views to supply missing features (e.g. commanding, events).

**Reply by ThibFrgsGmz** [reply]
The GDS will only have flight software engineers and system engineers as operators to verify the performance of the GNC algorithms and parameters/calibrations. 
The idea is to subject the software system to twisted cases to see how it responds, including using a digital twin.
Since these are laboratory tests, the longest duration would be for endurance testing, which could last several days.
System engineers would probably want to have tracking curves over several hours and coming in at 10 Hz (for example).
I mentioned a display of 30 Hz or more, but in reality it will be lower.


However, production operators may want to have it. If they want it too, the GUI will have to meet their needs, but that's going way too far and that's not the point.
This is just a demonstrator, so we have to limit ourselves to what we really need.
If production complains, we'll have to take a closer look and possibly modify the technology to suit their needs.



To have a consistent data history, the easiest way would be to save the data on the fly in a sqlite3 database for example?

**Reply by LeStarch** [reply]
In order to have persistent history, you could implement a history object that uses a database backend.  https://github.com/fprime-community/fprime-gds/blob/devel/src/fprime_gds/common/history/history.py  The default implementation is a RAM history (big list in python), however; you could implement an API for your database of choice.

**Reply by LeStarch** [reply]
Also, your willingness to help contribute is awesome.  Thank you!

**Reply by ThibFrgsGmz** [reply]
Thank you, but the game is not yet won.
I try to support your openness and excellent work, but I still have a lot of work to do, including the arduous task of changing mindsets. 
I'll probably still have questions by then!
